### PR TITLE
fix(BTLoss): ensure invariance to affine transformations

### PR DESCRIPTION
**Pull request by adosar**
The Barlow Twins loss must be invariant to affine transformations since ``corr(X, X) = corr(X, a*X + b)``.

Previously, the loss was not invariant to such transformations since during normalization the statistics from ``z_a`` and ``z_b`` were mixed up:

```python
combined = torch.stack([z_a, z_b], dim=0)  # Shape: 2 x N x D

normalized = F.batch_norm(
    combined.flatten(0, 1),  # Embeddings are not normalized independently.
    ...
)
```

You can check that the loss was not scale invariant with this snippet:
```python
import torch
from lightly.loss import BarlowTwinsLoss

torch.manual_seed(42)

loss = BarlowTwinsLoss()
x = torch.randn(32, 128)

print(loss(x, x))
print(loss(x, 2*x + 4))
```

Additional changes:
* Add unit test

See also:
* https://www.alphaxiv.org/abs/2103.03230v3?cid=677730ab31430e4d1bbf0f8a
* https://www.reddit.com/r/MachineLearning/comments/1htpuuv/r_how_barlow_twins_avoid_embeddings_that_differ/

### CODE REVIEW

File: lightly/loss/barlow_twins_loss.py [hunk 1]
```diff
@@ -87,24 +87,16 @@ def forward(self, z_a: torch.Tensor, z_b: torch.Tensor) -> torch.Tensor:
...
-) -> Tuple[torch.Tensor, torch.Tensor]:
-    """Helper function to normalize tensors along the batch dimension."""
-    # Stack tensors along a new dimension
-    combined = torch.stack([z_a, z_b], dim=0)  # Shape: 2 x N x D
-
-    # Normalize the stacked tensors along the batch dimension
-    normalized = F.batch_norm(
-        combined.flatten(0, 1),
+def _normalize(z: torch.Tensor) -> torch.Tensor:
+    """Helper function to normalize tensor along the batch dimension."""
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/loss/barlow_twins_loss.py:91]: Looks like it was indeed wrong and this fixes it. Same conclusion @guarin?

@adosar Thanks for the suggested change! Hope to be able to merge this soon. Could you address the failing formatting checks? You can simply run `make format` and it should run the formatting / import sorting. :) After that most of the CI issues should be also solved.

[the remark this answers is not in the document]
Comment (adosar) [reply lightly/loss/barlow_twins_loss.py:91]: Hi @liopeer,

Apologize for the formatting, I will fix it asap.

With regards to the [paper](https://arxiv.org/pdf/2103.03230):

> To simplify notations, $Z^A$ and $Z^B$ are assumed to be mean-centered along the batch dimension, such that each unit has mean output 0 over the batch.

Based on Equation 2 zero-mean but not necessarily unit standard deviation embeddings are enough for properly calculating the correlation matrix, since the denominator divides by standard deviation. 

This is of course equivalent to `bn(z_A).T @ bn(z_B)` as implemented in `lightly`, so IMO adding  `create batches of mean 0 and std 1` to the docstring seems good, as it precisely describes what happens under the hood.

## Update unit tests?

I was checking also the unit tests in `test_barlow_twins_loss` because I couldn't understand why the tests pass (before this PR, where normalization was mixed up) and I think `torch.allclose` should have stricter criteria (see the example below).

```python
# You need to run this script from project root.

from tests.loss.test_barlow_twins_loss import BarlowTwinsLossReference
from lightly.loss.barlow_twins_loss import BarlowTwinsLoss
import torch

torch.manual_seed(42)
 
batch_size, projector_size  = 32, 8912

x = torch.randn(batch_size, projector_size)
y = torch.randn(batch_size, projector_size)

loss = BarlowTwinsLoss()
loss_ref = BarlowTwinsLossReference(projector_size)

print(loss(x, y))
print(loss_ref(x, y))
```

https://github.com/lightly-ai/lightly/blob/477d3ceb8fa63bb8f79ed8c15b89bccedb0ad4f7/tests/loss/test_barlow_twins_loss.py#L83C1-L87C76

[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/loss/barlow_twins_loss.py:91]: Hi @adosar !
Sorry for the late reply, I was off for the last few days. Thanks a lot btw for above code, really convinced me that something is wrong & that your code gives the same results as the reference implementation. I will try to fix the last few things to make CI work & will commit my changes to the docstring. After that we should be ready to merge!

**Review by liopeer — APPROVED** [review]
Thanks for your contribution @adosar , it is much appreciated!

File: lightly/loss/barlow_twins_loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adosar) [reply lightly/loss/barlow_twins_loss.py]: Hi @liopeer,

 Could we also update the `rtol` and `atol` values in `test__loss_matches_reference_loss`? Because as shown in the above snippet, this was the reason the test (incorrectly) passed.

[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/loss/barlow_twins_loss.py]: You are right, I will do this in a follow-up PR. We should remove the values and use the torch defaults. Since the implementation is now correct we should only possibly have very small numerical differences.

[the remark this answers is not in the document]
Comment (adosar) [reply lightly/loss/barlow_twins_loss.py]: Perfect!
